incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1880,36 +1879,55 @@ private void onDecodePublish(
-                .typeId(mqttTypeId)
-                .publish(publishBuilder ->
-                {
-                    publishBuilder.qos(mqttPublishHeaderRO.qos);
-                    publishBuilder.flags(mqttPublishHeaderRO.flags);
-                    publishBuilder.expiryInterval(mqttPublishHeaderRO.expiryInterval);
-                    publishBuilder.contentType(mqttPublishHeaderRO.contentType);
-                    publishBuilder.format(f -> f.set(mqttPublishHeaderRO.payloadFormat));
-                    publishBuilder.responseTopic(mqttPublishHeaderRO.responseTopic);
-                    publishBuilder.correlation(c -> c.bytes(mqttPublishHeaderRO.correlationData));
-                    final Array32FW<MqttUserPropertyFW> userProperties = userPropertiesRW.build();
-                    userProperties.forEach(c -> publishBuilder.propertiesItem(p -> p.key(c.key()).value(c.value())));
-                });
-
+            int reasonCode = SUCCESS;
+            if (mqttPublishHeaderRO.qos > maximumQos)
+            {
+                reasonCode = QOS_NOT_SUPPORTED;
+            }
+            else if (mqttPublishHeaderRO.retained && !retainAvailable(capabilities))
+            {
+                reasonCode = RETAIN_NOT_SUPPORTED;
+            }
 
-            final MqttDataExFW dataEx = builder.build();
-            if (stream != null)
+            if (reasonCode != SUCCESS)
             {
-                stream.doPublishData(traceId, reserved, payload, dataEx);
+                onDecodeError(traceId, authorization, reasonCode);
+                decoder = decodeIgnoreAll;
+            }
+            else
+            {
+                final int topicKey = topicKey(mqttPublishHeaderRO.topic);
+                MqttPublishStream stream = publishStreams.get(topicKey);
+
+                final MqttDataExFW.Builder builder = mqttPublishDataExRW.wrap(dataExtBuffer, 0, dataExtBuffer.capacity())
+                    .typeId(mqttTypeId)
+                    .publish(publishBuilder ->
+                    {
+                        publishBuilder.qos(mqttPublishHeaderRO.qos);
+                        publishBuilder.flags(mqttPublishHeaderRO.flags);
+                        publishBuilder.expiryInterval(mqttPublishHeaderRO.expiryInterval);
+                        publishBuilder.contentType(mqttPublishHeaderRO.contentType);
+                        publishBuilder.format(f -> f.set(mqttPublishHeaderRO.payloadFormat));
+                        publishBuilder.responseTopic(mqttPublishHeaderRO.responseTopic);
+                        publishBuilder.correlation(c -> c.bytes(mqttPublishHeaderRO.correlationData));
+                        final Array32FW<MqttUserPropertyFW> userProperties = userPropertiesRW.build();
+                        userProperties.forEach(c -> publishBuilder.propertiesItem(p -> p.key(c.key()).value(c.value())));
+                    });
```

Comment:
Rename `p` to `pi` (properties item), `publishBuilder` to `p` (lambda variable) and use fluent style of method chaining to set `qos`, `flags`, ... in a single statement as we usually do.